bazel: reduce work done by bazel-generate.sh

RaduBerinde opened the pull request:
This seems to work for me, perhaps someone else can try it in their environment? I'll also take a look at the log in CI and make sure nothing unexpected happens.

----
#### bazel: reduce work done by bazel-generate.sh

The logic for dev generate bazel/make bazel-generate lives in the
script `build/bazelutil/bazel-generate.sh`.

If `go.mod`, `go.sum`, and `DEPS.bzl` are unchanged relative to
master, `bazel run //pkg/cmd/mirror` and `generate-staticcheck` can be
skipped. Similarly, if all `BUILD.bazel / *.bzl` files are unchanged,
then generate-test-suites can be skipped as well.

This commit adds a function that verifies whether a set of files is
unchanged from the upstream branch and uses it to implement these
improvements. The function is designed to fail silently if git is not
available or other errors are hit.

Fixes #73355.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/73893)
<!-- Reviewable:end -->

RaduBerinde commented:
Temporarily adding a bogus commit that adds an import without updating `BUILD.bazel`, to make sure it still fails in CI.

rickystewart reviewed (changes requested):
This looks good! A couple small requests:

1. Can you please allow setting an environment variable to disable this logic? Adding a conditional like `if [ ! -z COCKROACH_BAZEL_FORCE_GENERATE ]; then return 1; fi` to the top of `files_unchanged_from_upstream` would do it, I think.
2. Update `build/teamcity-check-genfiles.sh` to pass in that environment variable when running `build/bazelutil/bazel-generate.sh` (via `BAZEL_SUPPORT_EXTRA_DOCKER_ARGS="-e COCKROACH_BAZEL_FORCE_GENERATE=1"`)
3. (Optional) Also update `dev` to take a command-line flag that, when passed, sets that environment variable before calling into the script. (We do this with the `--mirror` argument to `dev generate bazel`, so you can just copy what we've done for that.)

In doing all of the above, we 1) make sure that this logic is unaffected in CI and 2) give everyone an easy way to opt-out if this code turns out to be buggy in a way none of us noticed.

RaduBerinde committed:
bazel: reduce work done by bazel-generate.sh

The logic for dev generate bazel/make bazel-generate lives in the
script `build/bazelutil/bazel-generate.sh`.

If `go.mod`, `go.sum`, and `DEPS.bzl` are unchanged relative to
master, `bazel run //pkg/cmd/mirror` and `generate-staticcheck` can be
skipped. Similarly, if all `BUILD.bazel / *.bzl` files are unchanged,
then generate-test-suites can be skipped as well.

This commit adds a function that verifies whether a set of files is
unchanged from the upstream branch and uses it to implement these
improvements. The function is designed to fail silently if git is not
available or other errors are hit.

These optimizations are disabled on teamcity and can be disabled by
adding a `--force` flag to `dev generate`.

Fixes #73355.

Release note: None

RaduBerinde reviewed (commented):
Good ideas, done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73893#-:-Mr8W-F8-VJuIKGgAdBu:b-a93i7a)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/generated/redact_safe.md, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73893#-Mr8W1mE7JcoW0fQjQVJ:-Mr8W1mE7JcoW0fQjQVK:b-icsyq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd560803caff7959c0ebc629e49053b2c5547ac/docs/generated/redact_safe.md#L42)):*
> ```Markdown
> pkg/sql/sem/tree/table_ref.go | `ID`
> pkg/sql/sqlliveness/sqlliveness.go | `SessionID`
> pkg/storage/enginepb/mvcc3.go | `*MVCCStats`
> ```

Any idea why my change causes this diff? It happens with or without the `--force` flag. Perhaps because I'm now always passing `env`?


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
build: fix locale in generate_redact_safe.sh

I noticed a spurious diff in the output of this script (as part of
`dev generate`). I found this in the manpage for sort:

> *** WARNING *** The locale specified by the environment affects sort order.
> Set LC_ALL=C to get the traditional sort order that uses native byte values.

This change adds `LC_ALL=C` to the sort calls in the script.

Release note: None

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/73893)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[docs/generated/redact_safe.md, line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/73893#-Mr8W1mE7JcoW0fQjQVJ:-Mr8ZV2KDhNI-d3jz4OY:bsskian) ([raw file](https://github.com/cockroachdb/cockroach/blob/7dd560803caff7959c0ebc629e49053b2c5547ac/docs/generated/redact_safe.md#L42)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Any idea why my change causes this diff? It happens with or without the `--force` flag. Perhaps because I'm now always passing `env`?
</blockquote></details>

Added a commit that fixes this.


<!-- Sent from Reviewable.io -->

rickystewart commented on build/bazelutil/bazel-generate.sh:
```diff
@@ -2,14 +2,60 @@
+  fi
+
+  # Check if the files are unchanged.
+  DIFF=$(git diff --no-ext-diff --name-only $BASE "$@") || return 1
```
Nit: I think if you add a `--` before `"$@"`, that should make it clear to `git` that those arguments are supposed to be paths.

rickystewart reviewed (approved):
lgtm

RaduBerinde commented:
bors r+
